core/src/main/scala/chisel3/reflect/DataMirror.scala

```diff
@@ -243,12 +243,12 @@ object DataMirror {
     def iterator = {
       val myItems = collector.lift(d).map { x => (x -> path) }
       val deepChildrenItems = d match {
-        case a: Record =>
+        case a: Record if (!hasProbeTypeModifier(a)) =>
           a._elements.iterator.flatMap {
             case (fieldName, fieldData) =>
               collectMembersAndPaths(fieldData, s"$path.$fieldName")(collector)
           }
-        case a: Vec[_] =>
+        case a: Vec[_] if (!hasProbeTypeModifier(a)) =>
           a.elementsIterator.zipWithIndex.flatMap {
             case (fieldData, fieldIndex) =>
               collectMembersAndPaths(fieldData, s"$path($fieldIndex)")(collector)
```

Comment:
Why are these here? It seems to me that DataMirror should return probes as well and perhaps the onus is on the user providing a `collector` function that excludes probes if that is the behavior they want?

Reply:
This stops it from recursing /through/ probes, not excluding them, FWIW.

The `collector` you're describing is used in, e.g., `excludeProbes`, but the idea here is that `Probe`s are opaque in the sense that's being iterated over here -- they're basically `Element` in that sense, with an underlying probed type that itself may have children but the probe itself does not.

As-is no `collector` definition (AFAIK?) can use this (pre-this-PR) to provide the useful behavior of treating probes as the leaves of aggregates that they are.

All that said, what this method is """supposed""" to do independent of our immediate needs I can't speak for (such as it's expected to recurse in same way other things do), so if it's not directly needed sure we can drop it, if that seems better!

Reply:
> This stops it from recursing /through/ probes, not excluding them, FWIW.

Good point, thanks for pointing that out

> The collector you're describing is used in, e.g., excludeProbes, but the idea here is that Probes are opaque in the sense that's being iterated over here -- they're basically Element in that sense, with an underlying probed type that itself may have children but the probe itself does not.
>
> As-is no collector definition (AFAIK?) can use this (pre-this-PR) to provide the useful behavior of treating probes as the leaves of aggregates that they are.

This is a fair point, but I think it highlights a much larger issue (and one that we do currently run into with `OpaqueTypes` as well) that we don't really have a way to talk about "leaves" other than `Element`. I think the change you're getting at here makes a lot of sense, but I think it necessitates coming up with the new API for "leaf".

> All that said, what this method is """supposed""" to do independent of our immediate needs I can't speak for (such as it's expected to recurse in same way other things do), so if it's not directly needed sure we can drop it, if that seems better!

You've convinced me that this is the right thing to do for this PR. We probably need to prioritize finding a better way to discuss leaves though.

I want to note that `collectLeafMembers` above is probably broken as it is expressed as:
```
  def collectLeafMembers(d: Data): Seq[Data] =
    DataMirror.collectMembers(d) { case x: Element => x }.toVector
```
It will ignore Aggregate probes even though those are leaves. It's also extremely likely that anyone calling that method is immediately matching on `Element` though....

Reply:
:100: , agreed and interesting/good to know about `OpaqueType`.  Hopefully soon we can get into business of working on these things and making everything awesome :rocket: .

Good catch re:`collectLeafMembers`, thank you for looking / thinking about that!! I think you're right.